### PR TITLE
Add GeneratorParam<LoopLevel>

### PR DESCRIPTION
**Pull request by steven-johnson**
This essentially behaves the same as ScheduleParam<LoopLevel>, with some minor differences in Stub usage. This is intended as a sketch to demonstrate how ScheduleParam<> could be eliminated entirely (see #2552 and shouldn't land until discussion there is clear on whether this is desirable.

(Note that this is built on top of https://github.com/halide/Halide/pull/2556)

### CODE REVIEW

**Comment by steven-johnson** [comment]
That requires that we have a reliable, unique string representation for every Func, and I'm pretty sure that is definitely not the case right now; Funcs can be given duplicate names in the front-end and we arbitrarily uniquify them later. What would `func.var` refer to for this?

     Func func("func"), another_func("func");

**Comment by dsharletg** [comment]
Hmm, LoopLevelContents already stores its values as strings: https://github.com/halide/Halide/blob/b29a3cbf4f93946741a235f62e98cbf940866789/src/Schedule.cpp#L31

**Comment by abadams** [comment]
Func names are unique'd on construction. LoopLevel can't hold a pointer to an actual Func or scheduling creates cyclic references and you get leaks.

**Comment by steven-johnson** [comment]
I suppose we could allow for "func.var" notation as a LoopLevel syntax, with the caveat that it would hard-fail if the name is ambiguous.

**Comment by steven-johnson** [comment]
> Could we get rid of the special cases for LoopLevel params if we had a good reliable syntax for representing LoopLevel as a string? Maybe `func.var`?

After a bit of looking into it: yes, but IMHO we shouldn't just yet.
1. We could use this syntax and just fail if there is an ambiguity, but that's gonna be unsatisfying and frustrating since it means certain combinations might fail at compile time when used in unpredicted ways.
1. `func.var` doesn't compose well across multiple Generators if they compose with each other (e.g. how many will have `input.x`?); we'd need to uniquify them further, most obviously by prepending the generator name `generatorname.func.var`
1. We probably don't want to just allow all funcs and vars to be referenced this way, as it would mean that we expose internal details that other Generators would rely on; we could restrict to allowing only `Input<>` or `Ouptut<>` funcs to be referenced, or perhaps add a way to explicitly export a looplevel by name.
1. Finally, it's not clear to me that there is definitely a need for this yet, and I'm reluctant to add this (given the above concerns) unless/until we clearly need it. (This is especially ironic given that this PR is step 1 in removing the added-without-needing-it `ScheduleParam`...)

So, PTAL, but I'm inclined to say the special-casing is worthwhile for now, and better than trying to prematurely address the issues above.

File: src/Generator.h [hunk 1]
```diff
@@ -552,6 +554,56 @@ class GeneratorParam_MachineParams : public GeneratorParamImpl<T> {
...
+        // (Note that ordinarily this is a bad idea, since LoopLevels
+        // can be mutated later on; however, this method is only
+        // called by the Generator infrastructure, on LoopLevels that
+        // will never be mutated, so this is really just an elaborate way
+        // to avoid runtime assertions.)
+        LoopLevel copy;
+        copy.set(this->value());
+        copy.lock();
+        if (copy.is_inlined()) {
+            return "LoopLevel::inlined()";
```
Comment (dsharletg) [inline src/Generator.h]: This doesn't need to match the strings in set_from_string? (and below)

Comment (steven-johnson) [reply src/Generator.h]: Nope: this is used to generate C++ code in the stub, to fill in the C++ default values in the GeneratorParams struct.

**Pull request merged**